pkg/retriever/httpretriever.go

```diff
@@ -0,0 +1,199 @@
+package retriever
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"time"
+
+	"github.com/benbjohnson/clock"
+	"github.com/filecoin-project/go-state-types/big"
+	"github.com/filecoin-project/lassie/pkg/events"
+	"github.com/filecoin-project/lassie/pkg/types"
+	"github.com/ipfs/go-cid"
+	"github.com/ipld/go-car/v2"
+	"github.com/ipld/go-ipld-prime"
+	cidlink "github.com/ipld/go-ipld-prime/linking/cid"
+	"github.com/ipni/go-libipni/metadata"
+	"github.com/multiformats/go-multicodec"
+)
+
+var (
+	ErrHttpSelectorRequest = errors.New("HTTP retrieval for an explicit selector request")
+	ErrNoHttpForPeer       = errors.New("no HTTP url for peer")
+	ErrBadPathForRequest   = errors.New("bad path for request")
+)
+
+const HttpDefaultInitialWait time.Duration = 0
+
+var _ TransportProtocol = &ProtocolHttp{}
+
+type ProtocolHttp struct {
+	Client *http.Client
+
+	// customCompare is for testing only, and should be removed when we have more
+	// ways to compare candidates so we can control ordering deterministically in
+	// our tests.
+	customCompare func(a, b ComparableCandidate, mda, mdb metadata.Protocol) bool
+}
+
+// NewHttpRetriever makes a new CandidateRetriever for verified CAR HTTP
+// retrievals (transport-ipfs-gateway-http).
+func NewHttpRetriever(getStorageProviderTimeout GetStorageProviderTimeout, client *http.Client) types.CandidateRetriever {
+	return NewHttpRetrieverWithDeps(getStorageProviderTimeout, client, clock.New(), nil, HttpDefaultInitialWait, nil)
+}
+
+func NewHttpRetrieverWithDeps(
+	getStorageProviderTimeout GetStorageProviderTimeout,
+	client *http.Client,
+	clock clock.Clock,
+	awaitReceivedCandidates chan<- struct{},
+	initialPause time.Duration,
+	customCompare func(a, b ComparableCandidate, mda, mdb metadata.Protocol) bool,
```

Comment:
Yeah, so this is `NewHttpRetrieverWithDeps()` which is intended for instantiating for testing. `NewHttpRetriever()` is the normal path with defaults. We've been using this pattern in a few places, including `NewGraphsyncRetrieverWithConfig` and `NewBitswapRetrieverFromDeps`.

The `customCompare` is particularly shameful because it's only a testing option so we get to control candidate order. https://github.com/filecoin-project/lassie/issues/197 https://github.com/filecoin-project/lassie/issues/92 should make this go away. Also if we move logic into `Connect()` then we have something to measure (and control) at least, like we do with Graphsync.